bpo-31718: Raise ValueError instead of SystemError when calling methods of uninitialized io.IncrementalNewlineDecoder objects

orenmn opened the pull request:
In addition:
* Add tests to `test_io` to verify that indeed `ValueError` is raised instead of `SystemError`.
* While we are here, fix refleaks in ``io.IncrementalNewlineDecoder.__init__()``.

<!-- issue-number: bpo-31718 -->
https://bugs.python.org/issue31718
<!-- /issue-number -->

orenmn committed:
init commit

orenmn committed:
fix refleaks in IncrementalNewlineDecoder.__init__()

orenmn committed:
add a NEWS.d item

orenmn committed:
improve the refleaks fix

orenmn committed:
yet another improvement to the refleaks fix

serhiy-storchaka commented on Modules/_io/textio.c:
```diff
@@ -309,8 +311,9 @@ _PyIncrementalNewlineDecoder_decode(PyObject *myself,
-        PyErr_SetString(PyExc_ValueError,
-                        "IncrementalNewlineDecoder.__init__ not called");
+        PyErr_Format(PyExc_ValueError,
+                     "%.200s.__init__() not called",
```
Revert this change. The problem is that `IncrementalNewlineDecoder.__init__` not called. This can happen if subclass' `__init__` doesn't call `super().__init__`. Subclass's `__init__` can be called, but `self->decoder == NULL` if `IncrementalNewlineDecoder.__init__` not called.

serhiy-storchaka commented on Lib/test/test_io.py:
```diff
@@ -3508,7 +3508,13 @@ def _check(dec):
 
 class CIncrementalNewlineDecoderTest(IncrementalNewlineDecoderTest):
-    pass
+    def test_uninitialized(self):
```
Add cpython_only? This is an implementation detail, if implement the constructor as `__new__` rather than `__init__`, the following errors will be not raised.

serhiy-storchaka commented on Modules/_io/textio.c:
```diff
@@ -248,9 +248,11 @@ _io_IncrementalNewlineDecoder___init___impl(nldecoder_object *self,
                                             PyObject *errors)
 /*[clinic end generated code: output=fbd04d443e764ec2 input=89db6b19c6b126bf]*/
 {
+    Py_XDECREF(self->decoder);
```
There is a race condition, destructor of `self->decoder` can call `IncrementalNewlineDecoder.__init__` . Use Py_XSETREF().

orenmn committed:
fix the error messages and the refleaks fix. add @cpython_only

orenmn commented:
Sorry about messing up so much.

serhiy-storchaka commented:
LGTM. I'll merge if @benjaminp doesn't have comments.

benjaminp reviewed (commented):
Two minor things:
1. Most of the _io module classes abstract away the reinitialization logic with a macro `CHECK_INITIALIZED`. It would be nice to do that here.
2. Checking for initialization should be the first thing a method does. In this implementation, `.reset()` can partially reset the object before raising an exception.

orenmn committed:
add CHECK_INITIALIZED_DECODER() and use it in the beginning of methods

orenmn committed:
Merge branch 'master' into bpo31718-dont-raise-sys-err

orenmn committed:
make sure to incref first

orenmn committed:
add CHECK_INITIALIZED_DECODER() to reset() and to newlines getter

orenmn commented:
Note that i added an `ok` field to `nldecoder_object`. I would change `CHECK_INITIALIZED_DECODER()` to rely on the `decoder` field if you think it is too expensive.
Also, i added `CHECK_INITIALIZED_DECODER()` to the beginning of the getter of `newlines`. I would remove that if you think this is too much of a change to the behavior.

serhiy-storchaka commented:
Python implementation of `.reset()` can partially reset the object before raising an exception.

I don't understand why the `ok` field is added. `__init__()` can fail only if `PyUnicode_FromString("strict")`, but the `error` field is not used and could be removed (but see bpo-31722). I would set `error = PyUnicode_FromString("strict")` if `error == NULL` before changing any fields (if we want to keep an unused field).

orenmn commented:
@benjaminp @serhiy-storchaka so should the C implementation allow `reset()` to partially reset the object?

auvipy reviewed (changes requested):
rebase is needed
